src/MDOP.jl

```diff
@@ -87,3 +220,134 @@ function beta_statistic(Y::Dataset, s::Vector; τs = 0:50 , w::Int = 1)
+set to 2.
+[^Hegger1999]: Hegger, Rainer and Kantz, Holger (1999). [Improved false nearest neighbor method to detect determinism in time series data. Physical Review E 60, 4970](https://doi.org/10.1103/PhysRevE.60.4970).
+"""
+function fnn(NNdist::T,NNdistnew::T;r::Float64=2.0) where {T}
+    @assert length(NNdist) == length(NNdistnew) "Both input vectors need to store the same number of distances."
+
+    # convert array of arrays into simple vectors, since we only look at K=1
+    NN_old = zeros(length(NNdist))
+    NN_new = zeros(length(NNdistnew))
+    for i = 1:length(NNdist)
+        NN_old[i]=NNdist[i][1]
+        NN_new[i]=NNdistnew[i][1]
+    end
+
+    # first ratio
+    ratio1 = 1/r    # since we input only z-standardized data with unit
+    # construct statistic
+    ratio2 = NN_new./NN_old
+
+    cond1 = ratio2 .> r
+    cond2 = NN_old .< ratio1
+
+    fnns = sum(cond1.*cond2)
+    fnns2 = sum(NN_old .< ratio1)
+
+    # store fraction of valid nearest neighbors
+    if fnns2 == 0
+        FNN = NaN;
+    else
+        FNN = fnns/fnns2
+    end
+
+    return FNN
+end
+
+
+"""
+    estimate_maximum_delay(s::Vector; tw=1:50, samplesize::Float=1.0)) -> τ_max, L
+computes an upper bound for the search of optimal delays, when using `MDOP`
+[`MDOP`](@ref) or `beta_statistic` [`beta_statistic`](@ref).
+
+## Description
+The input time series `s` gets embedded with unit lag and increasing dimension,
+for dimensions (or time windows) `tw` (`RangeObject`).
+For each of such a time window the `L`-statistic from Uzal et al. [^Uzal2011]
+will be computed. `samplesize` determines the fraction of points to be
+considered in the computation of `L` (see [`uzal_cost`](@ref)). When this
+statistic reaches its global minimum the maximum delay value `τ_max` gets
+returned. When `s` is a multivariate `Dataset`, `τ_max` will becomputed for all
+time series of that Dataset and the maximum value will be returned. The returned
+`L`-statistic is a Dataset of size (length(tw)*size(s,2)).
+
+[^Nichkawde2013]: Nichkawde, Chetan (2013). [Optimal state-space reconstruction using derivatives on projected manifold. Physical Review E 87, 022905](https://doi.org/10.1103/PhysRevE.87.022905).
+[^Uzal2011]: Uzal, L. C., Grinblat, G. L., Verdes, P. F. (2011). [Optimal reconstruction of dynamical systems: A noise amplification approach. Physical Review E 84, 016223](https://doi.org/10.1103/PhysRevE.84.016223).
+"""
+function estimate_maximum_delay(s::Vector{Float64}; tw=1:50, samplesize::Float64=1.0)
+    L = zeros(Float64,length(tw),1)
+    # loop over time windows
+    cnt = 1
+    for i in tw
+        Y_act = embed(s,i,1)
+        L[cnt] = uzal_cost(Y_act; samplesize = samplesize)
+        cnt +=1
+    end
+    ~, τ_max = findmin(L)
+    return tw[τ_max], Dataset(L)
+end
+
+function estimate_maximum_delay(s::Dataset; tw=1:50, samplesize::Float64=1.0)
+    τs_max = zeros(Int,size(s,2))
+    Ls = zeros(Float64,length(tw),size(s,2))
```

Comment:
Here the situation is similar as above. First, don't use `Float64` in Ls, and use instead `eltype(s)`. This ensures type stability if a user uses Float32 type numbers as well. Next, again, because you make `Ls` a matrix, its size is not type-inferable. The size of a Dataset instead is type-inferrable. I'd recommend to change the function definition to:
```julia
estimate_maximum_delay(s::Dataset{D, T}; tw=1:50, samplesize::Real=1.0) where {D, T}
```
then, inside the source of the function you can use both `D` (the dataset dimensionality, i.e. `size(s, 2)`) and `T` (the dataset number type), when making `Ls`.

Reply:
Thanks for that! At least my assumption about the `@code_warntype` output was kind of right :)
Btw. very helpful with explaining the parametrization of types!